xcode/Subconscious/SubconsciousTests/Tests_SphereFS.swift

```diff
@@ -0,0 +1,496 @@
+            body: body
+        )
+        
+        _ = try sphere.save()
+        
+        try sphere.remove(slug: "foo")
+        
+        _ = try sphere.save()
+        
+        let slugs = try sphere.list()
+        
+        XCTAssertEqual(slugs.count, 1)
+        XCTAssertTrue(slugs.contains("bar"))
+        XCTAssertFalse(slugs.contains("foo"))
+    }
+    
+    func testSaveVersion() throws {
+        let base = UUID()
+        
+        let globalStoragePath = try createTmpDir(path: "\(base)/noosphere")
+            .path()
+        
+        let sphereStoragePath = try createTmpDir(path: "\(base)/sphere")
+            .path()
+        
+        let noosphere = try Noosphere(
+            globalStoragePath: globalStoragePath,
+            sphereStoragePath: sphereStoragePath
+        )
+        
+        let sphereReceipt = try noosphere.createSphere(ownerKeyName: "bob")
+        
+        let sphere = try SphereFS(
+            noosphere: noosphere,
+            identity: sphereReceipt.identity
+        )
+        
+        let versionA = try sphere.version()
+        
+        try sphere.write(
+            slug: "foo",
+            contentType: "text/subtext",
+            body: "Test".toData(encoding: .utf8)!
+        )
+        
+        let version = try sphere.save()
+        
+        let versionB = try sphere.version()
+        
+        XCTAssertNotEqual(
+            versionA,
+            versionB,
+            "Save updates version"
+        )
+        
+        XCTAssertEqual(
+            version,
+            versionB,
+            "Save returns current version"
+        )
+    }
+    
+    func testFailedSync() throws {
+        let base = UUID()
+        
+        let globalStoragePath = try createTmpDir(path: "\(base)/noosphere")
+            .path()
+        
+        let sphereStoragePath = try createTmpDir(path: "\(base)/sphere")
+            .path()
+        
+        let noosphere = try Noosphere(
+            globalStoragePath: globalStoragePath,
+            sphereStoragePath: sphereStoragePath,
+            gatewayURL: "http://fake-gateway.fake"
+        )
+        
+        let sphereReceipt = try noosphere.createSphere(ownerKeyName: "bob")
+        
+        let sphere = try SphereFS(
+            noosphere: noosphere,
+            identity: sphereReceipt.identity
+        )
+        
+        // Should fail
+        _ = try? sphere.sync()
+        
+        try sphere.write(
+            slug: "foo",
+            contentType: "text/subtext",
+            body: "Test".toData(encoding: .utf8)!
+        )
+        
+        try sphere.save()
+        
+        let foo = try sphere.read(slashlink: "/foo")
+        
+        // Should fail
+        _ = try? sphere.sync()
+        
+        XCTAssertEqual(
+            foo.body.toString(),
+            "Test",
+            "Read current version"
+        )
+    }
+    
+    func testWritesThenCloseThenReopenWithScopes() throws {
```

Comment:
This test succeeds.

Note I init each Noosphere class in a separate `do` block. `do` blocks create a scope, guaranteeing that Noosphere's `deinit` will be called when scope exits. `deinit` is where we deallocate the Noosphere pointer.